### PR TITLE
vendor: use github.com/klauspost/pgzip instead of compress/gzip

### PR DESCRIPTION
**Pull request by giuseppe**
from my tests, I've seen a net improvement of around 30% on the wall
clock time in decompressing layers.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by vrothberg** [comment]
Awesome, LGTM.

Just one request for the commit message. Could you add some info which additional packages need to be vendored by users of containers/image:
```
github.com/klauspost/pgzip v1.2.1
github.com/klauspost/compress v1.4.1
github.com/klauspost/cpuid v1.2.0
```

This will hopefully make vendoring containers/image easier.

**Comment by giuseppe** [comment]
let's not merge yet.  I will prepare a WIP PR for Skopeo that uses this PR (as well as the coming PR for containers/storage) so we can play better with it

**Comment by rhatdan** [comment]
Nice work @giuseppe

**Comment by giuseppe** [comment]
WIP PR for skopeo here: https://github.com/containers/skopeo/pull/572

**Comment by mrunalp** [comment]
I think we should put all this into a CRI-O PR and make sure the e2e pass :)

**Comment by giuseppe** [comment]
> I think we should put all this into a CRI-O PR and make sure the e2e pass :)

yes sir!  Done here: https://github.com/kubernetes-sigs/cri-o/pull/1963 :-)

**Comment by giuseppe** [comment]
it fails to build Skopeo as Skopeo is not yet vendoring "github.com/klauspost/pgzip"

**Comment by vrothberg** [comment]
LGTM (https://github.com/containers/skopeo/pull/572 passed)

**Comment by rhatdan** [comment]
LGTM, lets merge since this passes the CRI-O Tests.

**Comment by rhatdan** [comment]
@giuseppe tests are failing?

**Comment by vrothberg** [comment]
> @giuseppe tests are failing?

That's because the CI uses Skopeo for testing which doesn't yet vendor the pgzip library. However, https://github.com/containers/skopeo/pull/572 passes so we can safely merge this branch.

File: .travis.yml [hunk 1]
```diff
@@ -21,7 +21,7 @@ script: >
   -e TRAVIS=$TRAVIS -e TRAVIS_COMMIT_RANGE=$TRAVIS_COMMIT_RANGE
   -e TRAVIS_PULL_REQUEST=$TRAVIS_PULL_REQUEST -e TRAVIS_REPO_SLUG=$TRAVIS_REPO_SLUG
   -e TRAVIS_BRANCH=$TRAVIS_BRANCH -e TRAVIS_COMMIT=$TRAVIS_COMMIT
-  -e GOPATH=/gopath -e TRASH_CACHE=/gopath/.trashcache
+  -e GOPATH=/gopath -e TRASH_CACHE=/gopath/.trashcache -e HOME=/gopath
```
Comment (mtrmac) [inline .travis.yml]: Why is this necessary? It would be nice to at least mention this in the commit message.

File: vendor.conf [hunk 2]
```diff
@@ -1,5 +1,3 @@
-github.com/containers/image
-
```
Comment (mtrmac) [inline vendor.conf]: These two lines are not a self-vendoring, and intentional — IIRC `trash` prefers them to be there. Or has that changed?

File: copy/copy.go [hunk 3]
```diff
@@ -18,6 +17,7 @@ import (
 	"github.com/containers/image/signature"
 	"github.com/containers/image/transports"
 	"github.com/containers/image/types"
+	gzip "github.com/klauspost/pgzip"
```
Comment (mtrmac) [inline copy/copy.go]: (Non-blocking nit: I’d slightly prefer using the explicit `pgzip` name throughout the code, to make it obvious that this is not the standard-library implementation.)

Comment (giuseppe) [reply copy/copy.go]: fixed

File: vendor.conf (hunk 2 above)
Comment (giuseppe) [reply vendor.conf]: I've dropped this change

File: .travis.yml (hunk 1 above)
Comment (giuseppe) [reply .travis.yml]: dropped.  I was just trying to fix the CI but then I realized it is just the missing dependency in the Skopeo vendor.conf

**Comment by vrothberg** [comment]
LGTM

**Comment by vrothberg** [comment]
@giuseppe @rhatdan I think we can merge this one and start the usual vendor marathon in the tools.

**Comment by TomSweeneyRedHat** [comment]
@vrothberg tests aren't too happy  :^(

```
go build "-buildmode=pie" -ldflags "-X main.gitCommit=a51e38e60d34a595bde0df06f1d39853cc04a5de" -gcflags "" -tags "btrfs_noversion libdm_no_deferred_remove" -o skopeo ./cmd/skopeo
vendor/github.com/containers/image/pkg/compression/compression.go:9:2: cannot find package "github.com/klauspost/pgzip" in any of:
	/tmp/tmp.rSecZe78dh/src/github.com/containers/skopeo/vendor/github.com/klauspost/pgzip (vendor tree)
	/usr/local/go/src/github.com/klauspost/pgzip (from $GOROOT)
	/tmp/tmp.rSecZe78dh/src/github.com/klauspost/pgzip (from $GOPATH)
Makefile:83: recipe for target 'binary-local' failed
make[1]: *** [binary-local] Error 1
make[1]: Leaving directory '/tmp/tmp.rSecZe78dh/src/github.com/containers/skopeo'
Makefile:69: recipe for target 'test-skopeo' failed
make: *** [test-skopeo] Error 2
```

**Comment by vrothberg** [comment]
> @vrothberg tests aren't too happy :^(

That's because the CI uses Skopeo which does not yet have this change vendored. It's a bit of a chicken-egg problem for such changes in c/image. @giuseppe opened a PR at Skopeo to check whether it likes this one here: https://github.com/containers/skopeo/pull/572

As https://github.com/containers/skopeo/pull/572 is green, I think we can merge it.

**Comment by vrothberg** [comment]
Merging (agreed with @giuseppe).
